lib/bolt/pal.rb

```diff
@@ -607,11 +607,33 @@ def run_task(task_name, targets, params, executor, inventory, description = nil)
       end
     end
 
-    def run_plan(plan_name, params, executor = nil, inventory = nil, pdb_client = nil, applicator = nil)
+    # TODO: Can we use this function for BoltSpec, or do we need to keep the old one?
+    def run_plan(plan_name, params, executor, inventory = nil, pdb_client = nil, applicator = nil)
+      # Start the main plan and round robin in the plan compiler, so that
+      # backgrounded tasks can finish once the main plan exits
       in_plan_compiler(executor, inventory, pdb_client, applicator) do |compiler|
-        r = compiler.call_function('run_plan', plan_name, params.merge('_bolt_api_call' => true))
-        Bolt::PlanResult.from_pcore(r, 'success')
+        # Create a Fiber for the main plan. This will be run along with any
+        # other Fibers created during the plan run in the round_robin, with the
+        # main plan always taking precedence in being resumed.
+        fiber = Fiber.new do
+          r = compiler.call_function('run_plan', plan_name, params.merge('_bolt_api_call' => true))
+          Bolt::PlanResult.from_pcore(r, 'success')
+        rescue Bolt::Error => e
+          Bolt::PlanResult.new(e, 'failure')
+        end
+        # Give the main plan Fiber a name
+        name = "Main plan #{plan_name}"
```

Comment:
I think these messages are going to be more legible if we treat this more as a name than a description. The user probably knows what plan they called let's just call this `main`.